src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
+        // the buffer of this image is set to (srcBuffer+Offset) and the related region is set to cover the whole copy region (offset from 0)
+        {
+            auto inCpuImageRegionsDynArray = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::ICPUImage::SBufferCopy>>(1);
+            auto& inCpuImageRegion = inCpuImageRegionsDynArray->front();
+            inCpuImageRegion = {};
+            inCpuImageRegion.bufferOffset = 0u;
+            inCpuImageRegion.bufferRowLength = mainRegion.bufferRowLength;
+            inCpuImageRegion.bufferImageHeight = mainRegion.bufferImageHeight;
+            inCpuImageRegion.imageSubresource.aspectMask = mainRegion.imageSubresource.aspectMask;
+            inCpuImageRegion.imageSubresource.mipLevel = 0u;
+            inCpuImageRegion.imageSubresource.baseArrayLayer = 0u;
+            inCpuImageRegion.imageOffset.x = 0u;
+            inCpuImageRegion.imageOffset.y = 0u;
+            inCpuImageRegion.imageOffset.z = 0u;
+            inCpuImageRegion.imageExtent.width    = regionToCopyNext.imageExtent.width;
+            inCpuImageRegion.imageExtent.height   = regionToCopyNext.imageExtent.height;
+            inCpuImageRegion.imageExtent.depth    = regionToCopyNext.imageExtent.depth;
+            inCpuImageRegion.imageSubresource.layerCount = core::max(regionToCopyNext.imageSubresource.layerCount, 1u);
+
+            auto localImageOffset = core::vector4du32_SIMD(currentBlockInRow, currentRowInSlice, currentSliceInLayer, currentLayerInRegion);
+            uint64_t offsetInCPUBuffer = mainRegion.bufferOffset + core::dot(localImageOffset, srcBufferByteStrides)[0];
+            uint8_t* inCpuBufferPointer = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(srcBuffer->getPointer()) + offsetInCPUBuffer);
+            asset::ICPUImage::SCreationParams inCPUImageParams = dstImageParams;
+            inCPUImageParams.flags = asset::IImage::ECF_NONE; // Because we may want to write to first few layers of CUBEMAP (<6) but it's not valid to create an Cube ICPUImage with less that 6 layers.
+            inCPUImageParams.format = srcImageFormat;
+            inCPUImageParams.extent = regionToCopyNext.imageExtent;
+            inCPUImageParams.arrayLayers = regionToCopyNext.imageSubresource.layerCount;
+            inCPUImageParams.mipLevels = 1u;
+            inCPUImage = asset::ICPUImage::create(std::move(inCPUImageParams));
+            assert(inCPUImage);
+            core::smart_refctd_ptr<asset::ICPUBuffer> inCPUBuffer = core::make_smart_refctd_ptr< asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>, true> >(srcBuffer->getSize(), inCpuBufferPointer, core::adopt_memory);
+            inCPUImage->setBufferAndRegions(std::move(inCPUBuffer), inCpuImageRegionsDynArray);
+            assert(inCPUImage->getBuffer());
+            assert(inCPUImage->getRegions().size() > 0u);
+        }
```

Comment:
I think you could hoist that to the iterator's constructor

just make an `ICPUImage` with all the regions there, ready to be used.

Reply:
Ok good idea, I'll try

Reply:
I was mainly afraid of hitting assert about setBufferAndRegions being somehow incompatible with the big CPUImage. I'll try, shouldn't be a problem now that I think about it